bdns/dns.go

```diff
@@ -209,253 +212,139 @@ func (dnsClient *impl) exchangeOne(ctx context.Context, hostname string, qtype u
 		// and ensures that chosenServer can't be a bare port, e.g. ":1337"
 		chosenServerIP, _, err = net.SplitHostPort(chosenServer)
 		if err != nil {
-			return
+			return nil, "", err
 		}
 
 		go func() {
-			rsp, rtt, err := client.Exchange(m, chosenServer)
+			resp, rtt, err := c.exchanger.Exchange(req, chosenServer)
 			result := "failed"
-			if rsp != nil {
-				result = dns.RcodeToString[rsp.Rcode]
+			if resp != nil {
+				result = dns.RcodeToString[resp.Rcode]
 			}
 			if err != nil {
-				dnsClient.log.Infof("logDNSError chosenServer=[%s] hostname=[%s] queryType=[%s] err=[%s]",
-					chosenServer,
-					hostname,
-					qtypeStr,
-					err)
+				c.log.Infof("logDNSError chosenServer=[%s] hostname=[%s] queryType=[%s] err=[%s]", chosenServer, hostname, qtypeStr, err)
 			}
-			dnsClient.queryTime.With(prometheus.Labels{
+			c.queryTime.With(prometheus.Labels{
 				"qtype":    qtypeStr,
 				"result":   result,
 				"resolver": chosenServerIP,
 			}).Observe(rtt.Seconds())
-			ch <- dnsResp{m: rsp, err: err}
+			ch <- dnsRes{resp: resp, err: err}
 		}()
 		select {
 		case <-ctx.Done():
-			if ctx.Err() == context.DeadlineExceeded {
-				dnsClient.timeoutCounter.With(prometheus.Labels{
-					"qtype":    qtypeStr,
-					"type":     "deadline exceeded",
-					"resolver": chosenServerIP,
-					"isTLD":    isTLD(hostname),
-				}).Inc()
-			} else if ctx.Err() == context.Canceled {
-				dnsClient.timeoutCounter.With(prometheus.Labels{
-					"qtype":    qtypeStr,
-					"type":     "canceled",
-					"resolver": chosenServerIP,
-					"isTLD":    isTLD(hostname),
-				}).Inc()
-			} else {
-				dnsClient.timeoutCounter.With(prometheus.Labels{
-					"qtype":    qtypeStr,
-					"type":     "unknown",
-					"resolver": chosenServerIP,
-				}).Inc()
+			switch ctx.Err() {
+			case context.DeadlineExceeded:
+				result = "deadline exceeded"
+			case context.Canceled:
+				result = "canceled"
+			default:
+				result = "unknown"
 			}
-			err = ctx.Err()
-			return
+			c.timeoutCounter.With(prometheus.Labels{
+				"qtype":    qtypeStr,
+				"result":   result,
+				"resolver": chosenServerIP,
+				"isTLD":    fmt.Sprintf("%t", !strings.Contains(hostname, ".")),
+			}).Inc()
+			return nil, "", ctx.Err()
```

Comment:
The doc comment for the `Client` interface reads:
> The second return value of each method is the address of the resolver used to conduct the query, and should be populated even when returning an error.

Here, like below, we already know `chosenServer` and should return it.